Spigot-Server-Patches/0553-NamespacedKey-based-plugin-chunk-tickets.patch

```diff
@@ -0,0 +1,267 @@
++
++    @Override
++    public boolean addKeyedPluginChunkTicket(int x, int z, org.bukkit.NamespacedKey key, io.papermc.paper.TickingLevel level) {
++        Preconditions.checkArgument(key != null, "null key");
++        Preconditions.checkArgument(level != null, "null level");
++
++        int ticketLevel = io.papermc.paper.PaperTickingLevel.getTicketLevel(level);
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        if (chunkDistanceManager.addTicketAtLevel(TicketType.KEYED_PLUGIN_TICKET, new ChunkCoordIntPair(x, z), ticketLevel, key)) {
++            this.getChunkAt(x, z); // ensure loaded, copied from addPluginChunkTicket
++            return true;
++        }
++        return false;
++    }
++
++    @Override
++    public boolean removeKeyedPluginChunkTickets(int x, int z, org.bukkit.NamespacedKey key) {
++        boolean result = false;
++        Collection<io.papermc.paper.TickingLevel> levels = getKeyedPluginChunkTicketLevels(x, z, key);
++        for (io.papermc.paper.TickingLevel level : levels) {
++            result |= removeKeyedPluginChunkTicket(x, z, key, level);
++        }
++        return result;
++    }
++
++    @Override
++    public boolean removeKeyedPluginChunkTicket(int x, int z, org.bukkit.NamespacedKey key, io.papermc.paper.TickingLevel level) {
++        Preconditions.checkArgument(key != null, "null key");
++        Preconditions.checkArgument(level != null, "null level");
++
++        int ticketLevel = io.papermc.paper.PaperTickingLevel.getTicketLevel(level);
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        return chunkDistanceManager.removeTicketAtLevel(TicketType.KEYED_PLUGIN_TICKET, new ChunkCoordIntPair(x, z), ticketLevel, key);
++    }
++
++    @Override
++    public void removeKeyedPluginChunkTickets(org.bukkit.NamespacedKey key) {
++        Preconditions.checkArgument(key != null, "null key");
++
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        chunkDistanceManager.removeIfAllTicketsFor(TicketType.KEYED_PLUGIN_TICKET,
++                ticket -> key.equals(ticket.getObjectReason()));
++    }
++
++    @Override
++    public void removeKeyedPluginChunkTickets(Plugin plugin) {
++        Preconditions.checkArgument(plugin != null, "null plugin");
++
++        String namespace = new org.bukkit.NamespacedKey(plugin, "ignored").getNamespace();
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        chunkDistanceManager.removeIfAllTicketsFor(TicketType.KEYED_PLUGIN_TICKET,
++                ticket -> namespace.equals(ticket.getObjectReason().getNamespace()));
++    }
++
++    @Override
++    public Collection<io.papermc.paper.TickingLevel> getKeyedPluginChunkTicketLevels(int x, int z, org.bukkit.NamespacedKey key) {
```

Comment:
This feels dirty. I'm not sure there's really any benefit to allowing to add the same ticket at multiple levels and return a collection of levels.

Lets move the add call to remove any existing ticket so that there can only be a single level, and this can be getKeyedPluginChunkTickingLevel(x, z, key);

note I said 'ticking level' and not 'ticket level'